src/beat_number.rs

```diff
@@ -1,90 +1,79 @@
-#[derive(PartialEq, Debug)]
+#[derive(PartialEq, Debug, Clone, Copy)]
 pub struct BeatNumber {
-    pub quarter_note: u32,
     pub sixteenth_note: u32,
 }
 
 impl BeatNumber {
     pub fn is_beginning_of_measure(&self) -> bool {
-        self.quarter_note == 1 && self.sixteenth_note == 1
+        self.sixteenth_note == 0
     }
 
     pub fn minus_sixteenths(&self, num_sixteenths: u32) -> BeatNumber {
-        if num_sixteenths < self.sixteenth_note {
-            return BeatNumber {
-                quarter_note: self.quarter_note,
-                sixteenth_note: self.sixteenth_note - num_sixteenths,
-            };
+        BeatNumber {
+            sixteenth_note: (self.sixteenth_note as i32 - num_sixteenths as i32) as u32 % 16,
         }
-        let num_quarter_notes = (num_sixteenths + 3) / 4;
+    }
+
+    pub fn add_sixteenths(&self, num_sixteenths: u32) -> BeatNumber {
         BeatNumber {
-            quarter_note: (self.quarter_note as i32 - num_quarter_notes as i32 - 1) as u32 % 4 + 1,
-            sixteenth_note: (self.sixteenth_note as i32 - num_sixteenths as i32 - 1) as u32 % 4 + 1,
+            sixteenth_note: (self.sixteenth_note + num_sixteenths) % 16,
         }
     }
+
+    pub fn duration_since(&self, other: &BeatNumber) -> u32 {
+        (self.sixteenth_note + 16 - other.sixteenth_note) % 16
```

Comment:
This assumes we're within a single measure since the other beat (otherwise we'd need to start tracking measure #'s I guess)